docs/KERNEL-DUMPS.md

```diff
@@ -0,0 +1,48 @@
+
+In the EVE environment, the system kernel and the dump-capture kernel are the same. Since the EVE kernel is quite minimal and contains all the options needed for a dump-capture kernel, a single compiled kernel binary simplifies the entire build process and does not require additional disk space to store the capture kernel.
+
+### Setup dump-capture kernel in EVE-OS
+
+First, memory must be reserved for dump capture kernel. The EVE-OS grub configuration has a kernel parameter `crashkernel=128M` which is only specified for the x86_64 architecture. After the system kernel is booted, the `/etc/init.d/000-kexec` script loads the same kernel binary into the reserved area by calling the kexec system tool.
+
+`kexec` loads the dump-capture kernel reusing the same command line with a couple of important differences: `irqpoll nr_cpus=1 reset_devices`, where:
+
+* The `irqpoll` boot parameter reduces driver initialization failures due to shared interrupts in the dump-capture kernel.
+
+* We generally don't have to bring up a SMP kernel just to capture the dump, hence we specify `nr_cpus=1` option while loading dump-capture kernel to restrict number of CPUs to 1 and to save memory.
+
+* `reset_devices` force drivers to reset the underlying device during initialization.
+
+* `nomodule` disable module load: make system simple and prevent possible further crash in the dump-capture kernel
+
+Once `kexec` is called the system kernel is ready to handle a system crash by jumping to a dump-capture kernel in case of a panic.
+
+### EVE-OS behaviour in case of a system crash (kernel panic)
+
+After successfully loading the dump-capture kernel as previously described in the introduction, the system will reboot into the dump-capture kernel if a system crash is triggered.  Trigger points are located in panic(), die(), die_nmi() and in the sysrq handler (ALT-SysRq-c). After the dump-capture kernel is booted, the file `/proc/vmcore` is used to get access to the memory of the crashed system kernel.
+
+EVE-OS has a `kdump` container that checks for `/proc/vmcore` and generates a minimal crash dump by calling the makedumpfile tool. The `kdump` container is part of the linuxkit onboot process and is run strictly after the storage is initialized. The minimal kernel dump collection includes the following steps:
+
+* Minimal kernel dump will be generated in the `/persist/kcrashes` folder.
+* Only 5 fresh kernel dumps is stored in the folder, old dumps are deleted.
+* Kernel panic message from the crashed system buffer is redirected to the tty console in order to show it on the connected monitor for debug purposes.
```

Comment:
Do there also get logged when the normal kernel boots again?

Reply:
This is only to avoid the nasty behavior of logging from the containers, which buffer the whole output, so you won't see a line once the script is not yet exited. The idea is to log the original dmesg and show it on a screen, possibly someone can take a look on the monitor. Easiest way is to redirect to kmsg, instead of dealing with containers linuxkit and all this ...

So normal kernel boot is not affected.